File: frontend/src/services/registration-invitations.ts
import type {
  RegistrationInvitation,
  RegistrationInvitationCreate,
  RegistrationInvitationCreateResponse,
  RegistrationInvitationsResponse,
} from "../domain/registration-invitations";
import { authenticatedFetch } from "./auth";

const apiBaseUrl = import.meta.env.VITE_API_BASE_URL ?? "";
const basePath = `${apiBaseUrl}/api/v1/registration-invitations`;

async function failure(response: Response) {
  const body = await response.json().catch(() => null);
  return new Error(
    typeof body?.detail === "string" ? body.detail : `Request failed (${response.status})`,
  );
}

export async function listRegistrationInvitations(fetcher = authenticatedFetch) {
  const response = await fetcher(basePath);
  if (!response.ok) throw await failure(response);
  return response.json() as Promise<RegistrationInvitationsResponse>;
}

export async function createRegistrationInvitation(
  payload: RegistrationInvitationCreate,
  fetcher = authenticatedFetch,
) {
  const response = await fetcher(basePath, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw await failure(response);
  return response.json() as Promise<RegistrationInvitationCreateResponse>;
}

export async function revokeRegistrationInvitation(
  identifier: string,
  fetcher = authenticatedFetch,
) {
  const response = await fetcher(`${basePath}/${encodeURIComponent(identifier)}`, {
    method: "DELETE",
  });
  if (!response.ok) throw await failure(response);
  return response.json() as Promise<RegistrationInvitation>;
}

File: backend/app/open_node/api/routes/registration_invitations.py
from typing import Annotated
from urllib.parse import quote
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Request, status

from open_node.api.dependencies import get_inventory_store
from open_node.domain.registration_invitations import (
    RegistrationInvitationCreate,
    RegistrationInvitationCreateResponse,
    RegistrationInvitationRead,
    RegistrationInvitationsResponse,
)
from open_node.services.inventory import InventoryStore, SubscriptionPlanNotFoundError
from open_node.services.registration_invitations import (
    RegistrationInvitationConflict,
    RegistrationInvitationUnavailable,
)

router = APIRouter(prefix="/registration-invitations", tags=["registration invitations"])


@router.get("", response_model=RegistrationInvitationsResponse)
def list_registration_invitations(
    store: Annotated[InventoryStore, Depends(get_inventory_store)],
):
    return store._registration_invitations().list()


@router.post(
    "", response_model=RegistrationInvitationCreateResponse, status_code=status.HTTP_201_CREATED
)
def create_registration_invitation(
    payload: RegistrationInvitationCreate,
    request: Request,
    store: Annotated[InventoryStore, Depends(get_inventory_store)],
):
    try:
        issued = store._registration_invitations().create(payload)
    except SubscriptionPlanNotFoundError as exc:
        raise HTTPException(404, str(exc)) from exc
    base = str(request.base_url).rstrip("/")
    return RegistrationInvitationCreateResponse(
        invitation=issued.invitation,
        registration_url=f"{base}/account#invite={quote(issued.token, safe='')}",
    )


@router.delete("/{identifier}", response_model=RegistrationInvitationRead)
def revoke_registration_invitation(
    identifier: UUID,
    store: Annotated[InventoryStore, Depends(get_inventory_store)],
):
    try:
        return store._registration_invitations().revoke(identifier)
    except RegistrationInvitationUnavailable as exc:
        raise HTTPException(404, str(exc)) from exc
    except RegistrationInvitationConflict as exc:
        raise HTTPException(409, str(exc)) from exc
